src/device/intrinsics/synchronization.jl

```diff
@@ -10,8 +10,15 @@ export threadfence, threadfence_block, threadfence_system
 Waits until all threads in the thread block have reached this point and all global and
 shared memory accesses made by these threads prior to `sync_threads()` are visible to all
 threads in the block.
+
+!!! note
+
+    CUDA.jl's behavior slightly differs from CUDA C here: This barrier is allowed to be
+    resolved from divergent branches, i.e., the barrier is not aligned and does not require
+    all threads in the block to execute the same barrier instruction. This is necessary
+    because the Julia compiler can introduce divergent branches, e.g., when union-splitting.
 """
-@inline sync_threads() = ccall("llvm.nvvm.barrier0", llvmcall, Cvoid, ())
+@inline sync_threads(id::Int=0) = ccall("llvm.nvvm.barrier.sync", llvmcall, Cvoid, (Int32,), id)
```

Comment:
Maybe you can just _add_ `sync_threads(id::Int)` and keep the old `sync_threads()`? People can then start using `sync_threads(0)` to see if it works well in practice.

Also, I don't mind just vendoring this definition and experimenting with it. (If it's only me complaining this thing.)

Reply:
You can also do an aligned barrier with a name. So it might be confusing that the named one is unaligned, while the unnamed one is still aligned.

Reply:
@tkf Did you end up vendoring this definition then?

Reply:
This hasn't come up as a problem and I had no chance to play with it.